main/src/library/Array.flix

```diff
@@ -57,10 +57,1390 @@ namespace Array {
+            }
+        }
+
+    ///
+    /// Returns `true` if and only if at least one element in `a` satisfies the predicate `f`.
+    ///
+    /// Returns `false` if `a` is empty.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def exists(f: a -> Bool, a: Array[a]): Bool =
+        existsHelper(f, a, length(a), 0) as & Pure
+
+    ///
+    /// Helper function for `exists`.
+    ///
+    def existsHelper(f: a -> Bool, a: Array[a], len: Int32, i: Int32): Bool & Impure =
+        if (i >= len)
+            false
+        else
+            if (f(a[i]) == true)
+                true
+            else
+                existsHelper(f, a, len, i+1)
+
+    ///
+    /// Returns `true` if and only if all elements in `a` satisfy the predicate `f`.
+    ///
+    /// Returns `true` if `a` is empty.
+    ///
+    pub def forall(f: a -> Bool, a: Array[a]): Bool =
+        forallHelper(f, a, length(a), 0) as & Pure
+
+    ///
+    /// Helper function for `forall`.
+    ///
+    def forallHelper(f: a -> Bool, a: Array[a], len: Int32, i: Int32): Bool & Impure=
+        if (i >= len)
+            true
+        else {
+            if (f(a[i]) == false)
+                false
+            else
+                forallHelper(f, a, len, i+1)
+        }
+
+    ///
+    /// Returns an array of every element in `a` that satisfies the predicate `f`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def filter(f: a -> Bool, a: Array[a]): Array[a] & Impure =
+        foldRight((x,xs) -> if (f(x)) x :: xs else xs, Nil, a) |> List.toArray
+
+    ///
+    /// Returns a pair of lists `(a1, a2)`.
+    ///
+    /// `a1` contains all elements of `a` that satisfy the predicate `f`.
+    /// `a2` contains all elements of `a` that do not satisfy the predicate `f`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def partition(f: a -> Bool, a: Array[a]): (Array[a], Array[a]) & Impure =
+        let step = { (x, ac) ->
+            let (a1,a2) = ac;
+            if (f(x)) (x :: a1, a2) else (a1, x :: a2)
+        };
+        let (xs,ys) = foldRight(step, (Nil,Nil), a);
+        (List.toArray(xs), List.toArray(ys))
+
+    ///
+    /// Returns a pair of arrays `(a1, a2)`.
+    ///
+    /// `a1` is the longest prefix of `a` that satisfies the predicate `f`.
+    /// `a2` is the remainder of `a`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def span(f: a -> Bool, a: Array[a]): (Array[a], Array[a]) & Impure=
+        match findIndexOfLeft(x -> !(f(x)), a) {
+            case None => (takeLeft(length(a), a), [])
+            case Some(i) => (takeLeft(i, a), dropLeft(i, a))
+        }
+
+    ///
+    /// Alias for `dropLeft`.
+    ///
+    pub def drop(n: Int32, a: Array[a]) : Array[a] & Impure = dropLeft(n, a)
+
+    ///
+    /// Returns a copy of array `a`, dropping the first `n` elements.
+    ///
+    /// Returns `[]` if `n > length(a)`.
+    ///
+    pub def dropLeft(n: Int32, a: Array[a]) : Array[a] & Impure =
```

Comment:
It is interesting that `dropLeft` has no impure equivalent -- because once an array has been created its size cannot be changed.